modules/cas_cache/layer_cache_management.c

```diff
@@ -1071,8 +1160,11 @@ int _cache_mngt_remove_core_prepare(ocf_cache_t cache, ocf_core_t core,
 	if (!cmd->force_no_flush) {
 		if (core_active) {
 			/* Flush core */
-			flush_result = _cache_mngt_core_flush_sync(core,
-					flush_interruptible);
+			if (flush_interruptible)
+				flush_result = _cache_mngt_core_flush_sync(core,
```

Comment:
What if the call gets interrupted and we exit this function and unlock & put the cache (in caller function) before async flush completes? Isn't this the exact scenario that you are attempting to fix ?

Reply:
I added second cache_get() and left read_unlock() only in flush completion